### PR TITLE
Make ScoreSortBuilder implement NamedWriteable and add fromXContent()

### PR DESCRIPTION
**Pull request by cbuescher**
This change makes ScoreSortBuilder implement NamedWriteable, adds equals() and hashCode() and also implements parsing ScoreSortBuilder back from xContent. This is needed for the ongoing Search refactoring.

### CODE REVIEW

**Comment by cbuescher** [comment]
@MaineC I took a shot at ScoreSortBuilder to get a first feeling for the sort builders, maybe you can take a look.

File: core/src/test/java/org/elasticsearch/search/sort/ScoreSortBuilderTests.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    /**
+     * test passing null to {@link ScoreSortBuilder#order(SortOrder)} is illegal
+     */
+    public void testIllegalOrder()  {
+        try {
+            new ScoreSortBuilder().order(null);
+            fail("NPE expected");
+        } catch (NullPointerException e) {
```
Comment (MaineC) [inline core/src/test/java/org/elasticsearch/search/sort/ScoreSortBuilderTests.java]: Not sure which junit version we are on. Came across the ExpectedException rule here https://github.com/junit-team/junit/wiki/Exception-testing recently. Just mentioning, no need to change this code right now as the above is the pattern we use all over the place.

**Comment by MaineC** [comment]
Only question: From briefly looking I couldn't find the code in fromXContent in the original source - was that newly added, or didn't I look close enough?

Other than that LGTM.

File: core/src/test/java/org/elasticsearch/search/sort/ScoreSortBuilderTests.java (hunk 1 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/sort/ScoreSortBuilderTests.java]: Looks great, I just tried it and it works for us. Will use it more often I think, better than `@Test(expected = ...)` which I think is a forbidden API now.

**Comment by cbuescher** [comment]
@MaineC thanks, great suggestion about ExpectedException rules. Regarding the code in fromXContent the original source, I think theres not one place where the "_score" sorts get parse but many. I think one of the main problems with the current sort refactoring is that the parsing code is so spread across SortParseElement. 
In this case, where SortField for score sort gets added in addSortField() the only parameter that matters is `reverse`. That in turn is only changed from the default in addCompoundSortField() when the `order` or `reverse` json parameter is specified (and for the short cut notation `{ "_score", "asc"}`, but I think we need to handle that separately). Hope this makes sense, would like to get your feedback on this since I think this is what makes the sort refactoring tricky.

**Comment by MaineC** [comment]
> I think one of the main problems with the current sort refactoring is that the parsing code is so spread 
> across SortParseElement. 

+1000 that's biggest issue that's been making things more complicated than they look like on first sight IMHO

> Hope this makes sense.

It does make sense.

Side-note: I think test-wise I'm really looking forward to something that compares the outcome of the current SortParseElement to what we will have for generating Lucene sort objects...

**Comment by cbuescher** [comment]
@MaineC thanks, pushed update including the proposed changes in the test an will merge if you don't have further comments.

**Pull request merged**